### PR TITLE
change(web): spins off the OSK component as its own subproject, converts it to use ES modules 🧩

### PR DESCRIPTION
**Pull request by jahorton**
This transforms the OSK code to allow it to stand alone, completely independent from the full Keyman Engine for Web as its own top-level project.  

There are additional changes in the code that refactor some subcomponents into lower-level modules and reorganizes some parts of them to facilitate a cleaner split; one of the goals is to prevent the OSK from needing to include the code used by keyboards for rule processing.  As such, Web's keyboard-processor module has itself been near-split in two - one part responsible for keyboard typing and layout metadata, with the other handling rule processing.  That project has not truly been split, but it's near-trivial to do after this PR's changes.

... yeah, this one's _**big**_.

To facilitate testing, a demo/test page has been added... though of course, our builds can't provide it yet because the main Web build-product is still broken at this point due to lack of ES-module conversion of the full codebase.

Preview of the test page:

image

### CODE REVIEW

File: android/KMEA/app/src/main/assets/android-host.js [hunk 1]
```diff
@@ -170,10 +170,8 @@ function deregisterModel(modelID) {
 function enableSuggestions(model, mayPredict, mayCorrect) {
   // Set the options first so that KMW's ModelManager can properly handle model enablement states
   // the moment we actually register the new model.
-  keyman.osk.banner.setOptions({
-    'mayPredict': mayPredict,
-    'mayCorrect': mayCorrect
-  });
+  keyman.core.languageProcessor.mayPredict = mayPredict;
```
Comment (darcywong00) [inline android/KMEA/app/src/main/assets/android-host.js:173]: Does this also apply to l.181?

Comment (jahorton) [reply android/KMEA/app/src/main/assets/android-host.js:173]: True, good catch.

File: web/src/engine/main/dom/utils.ts [hunk 2]
```diff
@@ -31,86 +31,86 @@ namespace com.keyman.dom {
...
-    /**
-     * Function     getAbsoluteX
-     * Scope        Public
-     * @param       {Object}    Pobj        HTML element
-     * @return      {number}
-     * Description  Returns x-coordinate of Pobj element absolute position with respect to page
-     */
-    static getAbsoluteX(Pobj: HTMLElement): number { // I1476 - Handle SELECT overlapping END
-      var Lobj: HTMLElement
+    // /**
```
Comment (mcdurdin) [inline web/src/engine/main/dom/utils.ts:34]: This commented code needs to be removed

File: web/src/engine/main/kmwbase.ts [hunk 3]
```diff
@@ -502,10 +502,35 @@ namespace com.keyman {
...
+
+      // Register a listener for model change events so that we can hot-swap the banner as needed.
+      // Handled here b/c banner changes may trigger a need to re-layout the OSK.
+
+      if(currentType != bannerController.activeType) {
+        this.refreshLayout();
+      }
+
+      return true;
+    });
```
Comment (mcdurdin) [inline web/src/engine/main/kmwbase.ts:532]: Indents look quite wonky here

File: web/src/engine/osk/src/keyboard-layout/oskLayerGroup.ts [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    // Set default OSK font size (Build 344, KMEW-90)
+    let layoutFS = layout['fontsize'];
+    if(typeof layoutFS == 'undefined' || layoutFS == null || layoutFS == '') {
+      ls.fontSize='1em';
+    } else {
+      ls.fontSize=layout['fontsize'];
+    }
+
+    ls.width = '100%';
+    ls.height = '100%';
```
Comment (mcdurdin) [inline web/src/engine/osk/src/keyboard-layout/oskLayerGroup.ts:36]: These seem new here?

**Review by mcdurdin — APPROVED** [review]
This is huge. I am unable to review this thoroughly as there are too many changes. So I think the only way forward is a full regression test -- and lots of ongoing testing throughout 17.0 alpha, with particular emphasis on edge cases and looking out for any change in behavior.

File: common/web/keyboard-processor/src/keyboards/activeLayout.ts [hunk 5]
```diff
@@ -149,6 +149,45 @@ export class ActiveKey implements LayoutKey {
...
   }
 
+  @Enumerable
+  public get baseKeyEvent(): KeyEvent {
+    return deepCopy(this._baseKeyEvent);
+  }
+
+  // Since I'll forget to note this otherwise... https://github.com/keymanapp/keyman/pull/7741/files#diff-9c3f288e2f41bf7a787f1aaff8802aee29c239f98d2efcb96d3a42a2ccfae0dd
+  // This implementation was apparently lost in a merge or something?  The question is... exactly where and when?  It's visible in
+  // the repo, but was missing from this branch... UNLIKE certain references to it!
```
[the remark this answers is not in the document]
Comment (jahorton) [reply common/web/keyboard-processor/src/keyboards/activeLayout.ts:159]: Yep.  The problem was... I wasn't anywhere near the state to make this into a PR at the time, so I worked with what I had.  Thanks for raising the comment - that's _exactly_ why I put it there, to be reminded of it here.

File: web/src/engine/osk/src/keyboard-layout/oskLayerGroup.ts (hunk 4 above)
Comment (jahorton) [reply web/src/engine/osk/src/keyboard-layout/oskLayerGroup.ts:36]: When developing the OSK module's demo-page, there was a case or two where omission of the newly-added styling resulted in a bad layout for the resulting OSK.  I'm not sure exactly what the problem case was (or "cases were") at this point, though.
